@W-18014214 useBlockNavigation implementation in SEO

jeremy-jung1 opened the pull request:
[W-18014214](https://gus.lightning.force.com/a07EE00002Avdo7YAB)

# Description

This is the implementation of the useBlockNavigation function in pwa-kit-react-sdk and its example usage with a temporary callback that executes a timeout. In [W-17530042](https://gus.lightning.force.com/a07EE000027V85TYAS) this callback will be replaced with logic that calls the getUrlMapping API and updates the routes. 

A change is also made in `extension-chakra-storefront` that adds a loading screen between the header and the footer while the navigation is blocked. 

# Types of Changes

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [x] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

# Changes

- Added extension-seo-url-mapping and . The name of this extension might still be up for debate
- Added loading screen to `extension-chakra-storefront`
- Implemented useBlockNavigation in `pwa-kit-react-sdk`

# How to Test-Drive This PR

## Testing URL redirects and non-URL redirects (Product and Category)
- (step1) Load this branch locally and do `npm start`
- (step2) Check for the URL type Redirect, which goes from Women's Clothing (Shop All) to Men's (Shop All). Click on Women's Clothing to see the redirect happen. See that there is a brief loading circle
- (step3) Check for the Category type Redirect, which goes from Women's Jewelry (Shop All) to Men's Accessories (Shop All). See that the URL is still for women's jewelry while the PLP is men's accessories because only the component has been loaded in.
- (step4) Check for the Product type Redirect, which goes from Men's Yartmouth Gloves (First in Men's Accessories) to Black Flat Front Wool Suit. 

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

jeremy-jung1 committed:
Initial commit

jeremy-jung1 committed:
Delete use-block-routing.ts

jeremy-jung1 committed:
type fix

jeremy-jung1 committed:
typescript minimal package.json

jeremy-jung1 commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -70,3 +71,39 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+ * @param {function} func
+ * @returns {boolean} State that is set to true during blocking, false otherwise.
+ */
+export const useBlockNavigation = (func) => {
```
Commenting to make it clearer what's been added: this hook will be added to pwa-kit-react-sdk

jeremy-jung1 commented on packages/extension-seo/src/contexts/index.tsx:
```diff
@@ -0,0 +1,55 @@
+    children: ReactNode
+}
+export const NavigationGuardContext = React.createContext<NavigationGuardContextType> ({isBlocked: false})
+export const NavigationGuardProvider: React.FC<NavigationGuardProviderProps> = ({callback, children}) => {
```
Commenting to make it clearer what's been added: A provider for a context related to navigation guard (aka the function that blocks navigation for seo url mapping). 

I'd like the context to contain the `isBlocked` boolean to be used in extension-chakra-storefront to display a loading screen while the url mapping is resolving, like so: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2247/files#diff-2d466b58d990c6eacb062f892061f4c2abecbb202238aff8666965831bea808bR407

Wondering how this can be done between two different extensions

jeremy-jung1 commented on packages/extension-seo/src/components/sample-hoc.tsx:
```diff
@@ -0,0 +1,46 @@
+type SampleHOCProps = React.ComponentPropsWithoutRef<any>
+
+// Define the HOC function
+const sampleHOC = <P extends object>(WrappedComponent: React.ComponentType<P>) => {
```
Commenting to make it clear what's been added: an HOC that feeds the callback function in as a navigation guard function.

jeremy-jung1 committed:
trying to override navigation guard context

jeremy-jung1 committed:
Fix errors

jeremy-jung1 committed:
add extension-seo-url-mapping

jeremy-jung1 committed:
fix extension-seo-url-mapping and delete extension-seo

jeremy-jung1 committed:
Put SkipNavLink back in and comment out isBlocked usage

jeremy-jung1 committed:
comment lots of code in sample-hoc

jeremy-jung1 committed:
Put things back in for sample-hoc

jeremy-jung1 committed:
Put necessary commenting in with-layout and react-query config

jeremy-jung1 committed:
Updates

jeremy-jung1 committed:
Add back skipnavcontent

jeremy-jung1 committed:
Pass abort controller signal into callback

jeremy-jung1 committed:
Revert context changes

jeremy-jung1 committed:
Revert import in setup-app

jeremy-jung1 committed:
Update wording

jeremy-jung1 committed:
Update CHANGELOG.md

jeremy-jung1 committed:
Merge branch 'feature/extensibility-v2' into useBlock

jeremy-jung1 committed:
try updates on CI

jeremy-jung1 committed:
Remove irrelevant shell

jeremy-jung1 commented on packages/template-typescript-minimal/package.json:
```diff
@@ -20,13 +20,16 @@
   },
   "devDependencies": {
     "@chakra-ui/react": "^2.8.2",
+    "@chakra-ui/skip-nav": "^2.0.15",
```
In spite of deleting node_modules and package-lock.json in this package and `extension-chakra-storefront` and doing `npm install`, the error:
`Module not found: Error: Can't resolve '@chakra-ui/skip-nav' in '/Users/jangho.jung/Documents/pwa-kit/packages/extension-chakra-storefront/src/components/with-layout'`

kept showing up. Adding the package in here temporarily fixed it. Will need to see what's causing this issue

jeremy-jung1 commented on packages/extension-seo-url-mapping/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,50 @@
+} from '@salesforce/pwa-kit-extension-sdk/react'
+type SeoHOCProps = React.ComponentPropsWithoutRef<any>
+
+const seoHOC = <P extends object>(WrappedComponent: React.ComponentType<P>) => {
```
Commenting to make it clear what's been added: 

an seo HOC that has the callback function that will soon hold the getUrlMapping API call. Currently it is a timeout.

jeremy-jung1 commented on packages/extension-seo-url-mapping/src/types/config.ts:
```diff
@@ -0,0 +1,29 @@
+ */
+export type Config = Required<UserConfig>
+
+interface StoreSlice {
```
Commenting to make it clear what's been added:

An extension store for isBlocked and its state updater so that it can be accessed from the storefront extension

bendvc committed:
Update package-lock files

Lint and add ignore file

bendvc committed:
Fix typo

bendvc committed:
Align version numbers for new extension

bendvc committed:
Make seo extension public so it will work with create-app

jeremy-jung1 committed:
Update package name

jeremy-jung1 committed:
Create .eslintignore

jeremy-jung1 committed:
Remove unused import

jeremy-jung1 committed:
new package locks

jeremy-jung1 committed:
Disable some test things to see what it does for CI

jeremy-jung1 committed:
Try to list dependencies in test action

jeremy-jung1 commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,50 @@
+type SeoHOCProps = React.ComponentPropsWithoutRef<any>
+
+const seoHOC = <P extends object>(WrappedComponent: React.ComponentType<P>) => {
+    const SeoHOC: React.FC<P> = (props: SeoHOCProps) => {
```
Commenting to make it clear what's been added:

an seo HOC that has the callback function that will soon hold the getUrlMapping API call. Currently it is a timeout.

jeremy-jung1 commented on packages/extension-commerce-bm-seo/src/types/config.ts:
```diff
@@ -0,0 +1,29 @@
+
+interface StoreSlice {
+    isBlocked: boolean
+    setIsBlocked: (someBool: boolean) => void
```
Commenting to make it clear what's been added:

An extension store for isBlocked and its state updater so that it can be accessed from the storefront extension

jeremy-jung1 commented:
Linting of the mono repo passed for me locally but CI still fails on setup ubuntu/windows machine and linting. Will be looking into it

jeremy-jung1 committed:
Testing CI

jeremy-jung1 committed:
Testing CI

jeremy-jung1 committed:
Testing CI

jeremy-jung1 committed:
Testing CI

jeremy-jung1 committed:
Testing CI

jeremy-jung1 committed:
Fix same link click doing the load again

vmarta commented on packages/extension-chakra-storefront/src/components/with-layout/with-layout.tsx:
```diff
@@ -302,28 +308,40 @@ const withLayout = <P extends object>(WrappedComponent: React.ComponentType<P>)
-                                outline: 0
-                            }}
-                        >
+                        {!isBlocked ? (
```
How about we be a bit more verbose? So it's clearer at first glance what is being blocked.

```suggestion
                        {!isNavigationBlocked ? (
```

vmarta commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,50 @@
+            return false
+        })
+
+        useApplicationExtension('@salesforce/extension-commerce-bm-seo')
```
Not needed anymore?
```suggestion
```

vmarta commented on packages/extension-commerce-bm-seo/src/types/config.ts:
```diff
@@ -0,0 +1,29 @@
+
+// This is safe to delete if your extension does not use state. If you aren't using this, ensure you remove the
+// `withApplicationExtensionStore` usage below as well.
+export const sliceInitializer: SliceInitializer<StoreSlice> = (set) => ({
```
Since this is not a type, let's move this to another file. For example, the extension-starter has it defined in setup-app.ts file.

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -70,3 +71,45 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+    const [isBlocked, setIsBlocked] = useState(false)
+    const funcRef = useRef()
+    const abortControllerRef = useRef(new AbortController())
+    funcRef.current = func
```
We can pass in `func` like this as the initial value.
```suggestion
    const funcRef = useRef(func)
    const abortControllerRef = useRef(new AbortController())
```

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -70,3 +71,45 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+        const unblock = block((location, action) => {
+            if (location?.pathname !== lastLocation.current?.pathname && funcRef.current) {
+                lastLocation.current = location
+                abortControllerRef.current.abort()
```
The abort controller is created but not get used anywhere (e.g. it's not [passed to `fetch`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController#examples)). Can we double check that this abort controller is working as intended?

jeremy-jung1 replied:
The abort is used when the user clicks on second link while the first link is loading. This aims to cancel the loading of the first link and immediately and show the latest result. The signal from the abort controller is being passed into the callback function, inside of which the abort call is listened to to cancel the timeout and unblock. In the next ticket I'll have to see how this would work with the getUrlMapping API call, but it might be a similar strategy.

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -70,3 +71,45 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+                lastLocation.current = location
+                abortControllerRef.current.abort()
+                abortControllerRef.current = new AbortController()
+                ;(async () => {
```
Curious.. what's the reason for running this async code this way?

jeremy-jung1 replied:
When I was working in template-retail-react-app, I had to wrap the asynchronous code of this callback function this way, or else the order of execution in the blocking callback would get messed up. I remember trying different ways of wrapping this async code, but this was the only one that worked. 

It may be different now, I'll look to make this more intuitive if I can

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -70,3 +71,45 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+                        push(location)
+                    }
+                })()
+                return false
```
How is this `false` going to be used?

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -70,3 +71,45 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+                })()
+                return false
+            }
+        })
```
`useEffect` can return a cleanup function if needed. In our case, there is something to clean up. 

```suggestion
        })
        return () => unblock()
```

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -70,3 +71,45 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+        })
+    }, [location])
+
+    return isBlocked
```
What if we return an object like this instead?

Previously, if someone were to call `useBlockNavigation`, they're allowed to name their variable anything they want:

```js
const block = useBlockNavigation(...) // this is allowed
```
Did you want them to name their variable the same each time? If yes, then let's return an object like this. I think it makes for better DX.

```suggestion
    return {isBlocked}
```

jeremy-jung1 replied:
We would want to provide a name for this variable, so that it doesn't conflict with the name for the state in the extension store

bendvc commented on packages/template-typescript-minimal/package.json:
```diff
@@ -1,7 +1,6 @@
 {
   "name": "typescript-minimal",
   "version": "4.0.0-extensibility-preview.4",
-  "private": true,
```
Why is this project now public?

bendvc commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,50 @@
+                    }
+                }, 10000)
+
+                signal.addEventListener('abort', () => {
```
Have you played around with how you are going to pass the signal to the scapi url mapping call.. I'm guessing you might grab the commerce api via the useCommerceApi and get the seo client and call it directly, can you pass fetch options to it?

vmarta replied:
It looks like we cannot pass in the fetch options _after_ the commerce api clients have been created. You can only pass in the fetch options initially to the `CommerceApiProvider`.

jeremy-jung1 replied:
The signal ended up not being necessary. Clicking on a second link immediately after a first link renders the latest link fine.

bendvc commented on packages/extension-commerce-bm-seo/dev/README.md:
You can probably go ahead an remove this `dev` project. Do to certain limitations these projects don't work when in the context of the mono repo. I believe we have a ticket to address this in the future, but until then we can simplify your pr by removing it.

jeremy-jung1 committed:
Update lock files with default npm registry

jeremy-jung1 committed:
Removing dev folder

jeremy-jung1 committed:
Apply some feedback

jeremy-jung1 committed:
Make seo extension public

jeremy-jung1 committed:
Make typescript minimal private

jeremy-jung1 committed:
Apply more feedback

jeremy-jung1 committed:
console log cleanup

jeremy-jung1 committed:
Merging other branch

jeremy-jung1 committed:
Integrate useUrlMapping call

jeremy-jung1 committed:
Merge branch 'feature/extensibility-v2' into useBlock-copy

jeremy-jung1 committed:
Merge branch 'feature/extensibility-v2' into useBlock-copy

jeremy-jung1 committed:
Handle URL redirect

jeremy-jung1 committed:
Linting and bring back a disabled test

jeremy-jung1 committed:
Remove verbose

jeremy-jung1 committed:
Add dynamic URL mapping for non-URL resourceType

jeremy-jung1 committed:
Support product detail

jeremy-jung1 committed:
linting

jeremy-jung1 committed:
Clean up

cc-prodsec commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/88758359-b034-48a9-998b-99b8eb599e14?refs=494d91ee41a9c8bf79dd5d919b077caa77edbca6&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/88758359-b034-48a9-998b-99b8eb599e14/license?refs=494d91ee41a9c8bf79dd5d919b077caa77edbca6&source=prChecksComment)

jeremy-jung1 committed:
Comment

jeremy-jung1 committed:
Change name

jeremy-jung1 commented on packages/extension-commerce-bm-seo/src/setup-app.ts:
```diff
@@ -0,0 +1,116 @@
+     * Provide resourceTypeToComponentMap to the PWA-Kit extension, so
+     * the SEO HOC can map the correct component to the resource type after resolivng URL Mapping API.
+     */
+    getComponentMatchingConfig(): object {
```
There needs to be a way to pass resourceTypeToComponentMap in a way that SEO HOC can access it and render the correct component depending on the resource type from the url mapping API. Perhaps this mapping can be optionally defined per extension and be combined together.

jeremy-jung1 committed:
Update location of resourceTypeToComponentMap

jeremy-jung1 commented on packages/extension-chakra-storefront/src/setup-app.tsx:
```diff
@@ -131,6 +131,15 @@ class ChakraStorefront extends ApplicationExtension<Config> {
+     * Provide resourceTypeToComponentMap to the PWA-Kit extension, so
+     * the SEO HOC can map the correct component to the resource type after resolivng URL Mapping API.
+     */
+    getComponentMatchingConfig(): object {
```
There needs to be a way to pass resourceTypeToComponentMap in a way that SEO HOC can access it and render the correct component depending on the resource type from the url mapping API. This mapping can be optionally defined per extension and be combined together to be passed to SEO HOC via the routes context

jeremy-jung1 commented on packages/extension-chakra-storefront/src/components/with-layout/with-layout.tsx:
```diff
@@ -113,13 +115,20 @@ const withLayout = <P extends object>(WrappedComponent: React.ComponentType<P>)
         const location = useLocation()
         const authModal = useAuthModal()
+        const [siteLocale, _] = useState(siteConfig.locale)
+        const {isNavigationBlocked, setSiteLocale} = useApplicationExtensionsStore((state) => {
```
TODO: handle lack of SEO extension

jeremy-jung1 committed:
get rid of pwa kit react sdk changes

jeremy-jung1 committed:
Add newline

jeremy-jung1 commented on packages/extension-chakra-storefront/src/components/with-layout/with-layout.tsx:
```diff
@@ -313,28 +321,40 @@ const withLayout = <P extends object>(WrappedComponent: React.ComponentType<P>)
+                                    flex="1"
+                                >
+                                    <OfflineBoundary isOnline={false}>
+                                        <WrappedComponent {...(props as P)} />
```
This part seems to make it necessary, in the context of isNavigationBlocked, that the order of extensions in package.json of template typescript minimal be SEO extension first and then storefront later because otherwise the HOC code gets called in the `WrappedComponent` and the logic in there won't be called to toggle the `isNavigationBlocked` when it's out of render.

jeremy-jung1 replied:
@hajinsuha1 spoke separately that he was able to make the order with SEO extension first and then storefront work 👍

jeremy-jung1 committed:
Remove location in return

jeremy-jung1 committed:
Update default.json

jeremy-jung1 committed:
Merge branch 'feature/extensibility-v2' into useBlock

jeremy-jung1 committed:
Lint and handle lack of seo extension

bendvc commented on packages/extension-chakra-storefront/config/default.json:
```diff
@@ -186,5 +186,10 @@
         "showDefaults": true,
         "interpretPlusSignAsSpace": false
+    },
+    "resourceTypeToComponentMap": {
+        "category": "ProductList",
+        "product": "ProductDetail",
+        "content_asset": "ProductList"
```
I'd expect that the resourceType here is capitalized like the return value before in the [API docs](https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-seo?meta=getUrlMapping)

hajinsuha1 replied:
Interestingly the actual API returns the resource type in lower case
<img width="1062" alt="Screenshot 2025-05-01 at 10 58 02 AM" src="https://github.com/user-attachments/assets/85d7f6a8-6e16-41b0-9951-8f5937c9ade8" />

bendvc commented on packages/extension-commerce-bm-seo/.eslintrc.js:
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+module.exports = {
+    extends: [require.resolve('@salesforce/pwa-kit-dev/configs/eslint')]
+}
```
Was this file not automatically generated for you when you created the extension project?

jeremy-jung1 replied:
I believe it was, this PR still includes the new extension project starter files

bendvc commented on packages/extension-commerce-bm-seo/.npmignore:
```diff
@@ -0,0 +1 @@
+dev/
```
This is a good addition.. we should be sure to update the starter-extension so that it has this file already.

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,108 @@
+:loudspeaker: Hey there, Salesforce Commerce Cloud community!
```
Lets update this readme to be specific to this extension.

hajinsuha1 replied:
@jeremy-jung1 I've started an initial README.md in my PR (with the help of cursor):
https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2290/commits/a2113daab7d207881c77e78dbca2e50a88510bc8

hajinsuha1 replied:
We probably also want to note in the README that the extension has to come before other extensions so that the HOC works properly

bendvc commented on packages/extension-commerce-bm-seo/config/default.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "enabled": true,
+    "commerceApi": {
```
I think you'll want to add the default for the resource mapping here.

bendvc commented on packages/extension-commerce-bm-seo/package.json:
```diff
@@ -0,0 +1,46 @@
+{
```
You might want to update this with the latest version from the `sample-extension` project as it looks like this might have been generated after we made some changes, specifically to the meta part

jeremy-jung1 replied:
Hm what is `sample-extension` project? Would you be referring to the `extension-starter`? Looks like a separate package outside of the repo. Where can I find it?

vmarta commented on packages/extension-commerce-bm-seo/.eslintignore:
```diff
@@ -0,0 +1,5 @@
+build
```
The SEO extension is introduced in Jinsu's PR, right? Should this PR be merged to his PR instead (and not to the extensibility feature branch)?

jeremy-jung1 replied:
I'm not sure which one will get merged to `extensibility-v2` first, but I don't want his PR to get extremely large by setting his as the base branch (that will make reviewing hard too). We already have a change that went into the extensibility feature branch, and it's been way better to consume each other's work that way as we collaborate in parallel.

vmarta commented on packages/internal-lib-build/package-lock.json:
A lot of package lock files that got updated.. I think they made it difficult to see which files I should be looking at in this PR.

I know that at some point you ran into an issue, that's why. But I'm not sure that you really need to update all of the lock files. Typically, the lock files are updated only when the corresponding package.json files are updated.

jeremy-jung1 replied:
Yeah sorry about that, but I still have this issue though. I tried to clear node_modules and package-lock.json with only specific dependencies, but I can't do `npm i` on a dependency individually because there are circular dependencies (e.g. pwa-kit-dev <-> pwa-kit-extension-sdk) and for some reason clearing all package-lock.json and node_modules and installing it at the root was the only way so far to get things to work

vmarta replied:
@jeremy-jung1 🤔 re: `npm i`, yeah you can run it only at the monorepo root because Lerna would take over and link all things together so the installation is successful. So running it inside one of the packages will give you an error.

Here's what I typically do: cleaning the node_modules only and then `npm i` at the root. The goal is to _update_ the lock files, rather than replacing them.

You can try the following:
1. Restore every lock file to the copies found in the base branch
2. Then `npx lerna clean` at the root to delete all of the node_modules
3. Finally, `npm i`
4. You should now have updated those package lock files that really need updating

jeremy-jung1 replied:
Thanks!! I think that worked. Looks a lot better. I'll do something like this more often now

jeremy-jung1 committed:
README changes and starter-extension change

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -72,6 +73,36 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+            }
+        })
+        return () => unblock()
+    }, [location])
```
It turns out that `block` is meant to be called sparingly. It's meant to register a callback whenever a navigation change is about to happen. Like an onClick event handler, you call it once to register.

So then with `unblock`, you can read it as "unregister the block handler". And similarly, you call that once (in the returned cleanup function).

```suggestion
    useEffect(() => {
        const unblock = block(async (location, action) => {
            if (location?.pathname !== lastLocation.current?.pathname && funcRef.current) {
                lastLocation.current = location
                setIsBlocked(true)
                await funcRef.current(location, action)
                setIsBlocked(false)
                // unblock()
                push(location?.pathname)
            }
        })
        return () => unblock()
    }, []) // Make sure to register the navigation block handler _once_
```

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js:
```diff
@@ -72,6 +73,36 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
+ * The function must return false to unblock the navigation
+ *
+ * @param {function} func
+ * @returns {boolean} State that is set to true during blocking, false otherwise.
```
Let's fix this outdated return signature.

jeremy-jung1 committed:
Add tests and do linting

jeremy-jung1 committed:
Revert "README changes and starter-extension change"

This reverts commit 0dfe98ed511e981c9c358c11e1723265976fb058.

jeremy-jung1 committed:
Remove unused README and update README for seo extension

jeremy-jung1 commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,124 @@
+
+The extension uses the Commerce API's Shopper SEO URL Mapping endpoint to handle SEO-friendly URLs and ensure proper routing within your PWA Kit application.
+
+## Usage of `isNavigationBlocked` and `siteLocale`
```
Included README changes from @hajinsuha1  PR (Credits to him) so that I can add documentation about the extension store states on top

jeremy-jung1 commented on packages/extension-commerce-bm-seo/README.md:
````diff
@@ -0,0 +1,124 @@
+}
+```
+
+Please note that the this extension needs to be defined in the array before any other extension that uses `isNavigationBlocked` to
````
@hajinsuha1 Added note about the order of the seo extension here

bendvc replied:
This shouldn't be a requirement as previously discussed. So we might want to edit/remove this.

jeremy-jung1 committed:
Remove template typescript minimal changes

jeremy-jung1 committed:
Merge branch 'feature/extensibility-v2' into useBlock

jeremy-jung1 committed:
Add .npmignore to extension-starter

jeremy-jung1 commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.test.tsx:
```diff
@@ -0,0 +1,227 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import React from 'react'
```
Credits to @yunakim714 for some of the mocks here from her PR. I put it also here so that much of test file is the same to make resolving merge conflicts easier.

jeremy-jung1 committed:
package.json to align with extension-starter

jeremy-jung1 committed:
Update CHANGELOG.md

vmarta commented on packages/extension-commerce-bm-seo/src/setup-app.ts:
```diff
@@ -0,0 +1,88 @@
+    setIsNavigationBlocked: (newIsNavigationBlocked) =>
+        set((state) => ({...state, isNavigationBlocked: newIsNavigationBlocked})),
+    siteLocale: undefined,
+    setSiteLocale: (config) => set((state) => config)
```
This way of setting the state does not match with what's on line 44 above.

vmarta commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,128 @@
+                }
+            },
+            {
+                enabled: false
```
I think it's worth adding a comment as to why this is disabled initially.

vmarta commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,128 @@
+            }
+        )
+
+        useEffect(() => {
+            setIsNavigationBlocked(isNavigationBlocked)
+        }, [isNavigationBlocked])
```
Similarly, we can add a comment here to note that we're broadcasting to other components via this shared state.

vmarta commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,128 @@
+        )
+
+        useEffect(() => {
+            const fetchData = async () => {
+                if (urlSegment) {
+                    const result = await refetch()
+                    if (resolveRef.current) {
+                        if (result.status === 'error') {
+                            resolveRef.current(undefined)
+                        } else {
+                            if (result.data?.destinationUrl) {
+                                resolveRef.current(result.data)
+                            } else {
+                                resolveRef.current(undefined)
+                            }
+                        }
+                    }
+                }
+            }
```
To be more readable, it'd be great if we flatten the nested conditions. Something like this:

```suggestion
            const fetchData = async () => {
                if (!urlSegment || !resolveRef.current) {
                    return
                }
                const result = await refetch()

                if (result.status === 'error') {
                    resolveRef.current(undefined)
                    return
                }

                const hasDestinationUrl = result.data?.destinationUrl
                resolveRef.current(hasDestinationUrl ? result.data : undefined)
            }
```

jeremy-jung1 committed:
Update tests

jeremy-jung1 committed:
restore package-lock.json

jeremy-jung1 committed:
Update apply-application-extensions.test.ts

jeremy-jung1 committed:
Update setup-jest.js

jeremy-jung1 committed:
Apply feedback

jeremy-jung1 committed:
Merge branch 'feature/extensibility-v2' into useBlock

jeremy-jung1 committed:
remove console log

jeremy-jung1 committed:
Update dependencies in seo extension

jeremy-jung1 committed:
all package-lock json changes

jeremy-jung1 committed:
Revert "all package-lock json changes"

This reverts commit cf056591171cbce71346573b75ef32e4aeacc278.

jeremy-jung1 committed:
linting?

jeremy-jung1 committed:
Revert "linting?"

This reverts commit 60084b2157c41395d61df7a4435bbba12f5dc121.

jeremy-jung1 committed:
Fix test

jeremy-jung1 committed:
linting

jeremy-jung1 commented on packages/template-typescript-minimal/app/components/_app-config/index.js:
```diff
@@ -18,6 +18,7 @@ const options = {
             queries: {
                 retry: false,
                 staleTime: 2 * 1000,
+                refetchOnWindowFocus: false,
```
Making this config the same as [template retail react app](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/2bcbbc2ac5abce0113c8e055eab902ffbb7d66bf/packages/template-retail-react-app/app/components/_app-config/index.jsx#L119-L153), otherwise the chakra storefront page will continue to refresh when theres a pointer event.

jeremy-jung1 committed:
Try just npm ci

jeremy-jung1 committed:
Give extension name in setup app

jeremy-jung1 committed:
Try something

jeremy-jung1 committed:
template-typescript-minimal setup

jeremy-jung1 committed:
Revert "template-typescript-minimal setup"

This reverts commit ff15616e6abf3bd2e05276d7daf8d7218ae7e7c5.

jeremy-jung1 committed:
Try lerna bootstrap

jeremy-jung1 committed:
Change to npm i

jeremy-jung1 committed:
Some echos

jeremy-jung1 committed:
Trying building in bootstrap

bendvc commented on packages/extension-chakra-storefront/src/components/with-layout/with-layout.tsx:
```diff
@@ -313,28 +328,40 @@ const withLayout = <P extends object>(WrappedComponent: React.ComponentType<P>)
+                                    flex="1"
+                                >
+                                    <OfflineBoundary isOnline={false}>
+                                        <WrappedComponent {...(props as P)} />
```
As mentioned in our direct chat, we might want to use an overlay for the spinner as opposed to controlling what components are rendered. If we do this it means the application and its router and providers all get unmounted and remounted and I believe this is having some unwanted side effects, I think one of those is that the route context is being reset.

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,124 @@
+# SEO Extension
```
lets update this name to reflect the name choice in the package/project.

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,124 @@
+
+### Configuration Options
+
+- `commerceAPI`: Settings for connecting to the Commerce API
```
We probably want to mention that this configuration is optional. We might want to mention that if a commerce provider is already added from another extension, that instance will be used.

bendvc commented on packages/extension-commerce-bm-seo/README.md:
```diff
@@ -0,0 +1,124 @@
+
+- The SEO extension provides an `isNavigationBlocked` state via the application extension store.
+- Downstream extensions (such as `@salesforce/extension-chakra-storefront`) should read this value to determine whether navigation is being blocked (e.g., while waiting for SEO URL mapping to resolve).
+- When `isNavigationBlocked` is `true`, the downstream extension can use the state to display a loading state until the SEO extension has finished processing the URL mapping.
```
There is no requirement that extensions are "downstream". So we might want to use something like "sibling" or simply "other installed extensions"

bendvc commented on packages/extension-commerce-bm-seo/README.md:
````diff
@@ -0,0 +1,124 @@
+  // Show loading spinner or block navigation
+}
+```
+### siteLocale and setSiteLocale
+
+The SEO extension also provides a `setSiteLocale` function in the application extension store.
+`siteLocale` represents the current locale for the site and used in the SEO Url Mapping API to ensure that the SEO extension are kept in sync with the current locale. It is set and managed by downstream extensions (such as your storefront) using the `setSiteLocale` function.
+
+**Example usage**
+```typescript
+const {setSiteLocale} = useApplicationExtensionsStore((state) =>
+  state.state['@salesforce/extension-commerce-bm-seo'] || {setSiteLocale: () => {}}
+)
+useEffect(() => {
+  setSiteLocale(currentLocale) // currentLocale is determined by your app logic
+}, [currentLocale])
+```
````
We should have a discussion about this addition. Does the commerce sdk provider not have a hook that provides the locale that it's currently using?

jeremy-jung1 replied:
I was seeing commerceApi being defined for each extension in `package.json` so I thought there would be restrictions in the information sharing between extensions, but I see now that the config is the same. I'll remove this

bendvc commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,127 @@
+    destinationUrl?: string
+}
+
+const getComponent = (
+    routes: Array<{component: React.ComponentType<any> | undefined; path: string}>,
+    resourceTypeToComponentMap: {[key: string]: string},
+    resourceType: string
+) => {
+    const ComponentClass = routes.find((_route) =>
+        _route.component?.displayName?.includes(resourceTypeToComponentMap[resourceType])
+    )?.component
+    return ComponentClass
+}
```
Lets move this to a utils file or sorts. Also add some comments at to what this function does. We are trying to have better file organization and trying to keep 1 method/function per file when we can.

bendvc commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,127 @@
+import {useExtensionConfig} from '../hooks/use-extension-config'
+type SeoHOCProps = React.ComponentPropsWithoutRef<any>
+
+interface UrlMappingResponse {
+    resourceType?: string
+    resourceId?: string
+    destinationUrl?: string
+}
```
So we want to add this in a types file?

bendvc commented on packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx:
```diff
@@ -0,0 +1,127 @@
+            async (location: Location, _: string) => {
+                const urlMappingResponse = await new Promise<UrlMappingResponse | undefined>(
+                    (resolve, __) => {
+                        const nextSegment = location.pathname
+                        // So that this promise can be resolved and navigation is unblocked outside this function
+                        resolveRef.current = resolve
+                        setUrlSegment(nextSegment)
+                    }
+                )
+                // If no redirect rule exists, go to original url
+                if (urlMappingResponse === undefined) {
+                    return
+                }
+                let Component
+                let props
+                // If the Redirect type is URL do a Redirect, else load matching component
+                if (!urlMappingResponse.resourceType) {
+                    Component = Redirect
+                    props = {
+                        to: urlMappingResponse.destinationUrl
+                    }
+                } else {
+                    Component = getComponent(
+                        routes,
+                        resourceTypeToComponentMap,
+                        urlMappingResponse.resourceType
+                    )
+                    props = {
+                        [`${urlMappingResponse.resourceType}Id`]: urlMappingResponse.resourceId
+                    }
+                }
+                setRoutes([
+                    {
+                        path: location.pathname,
+                        component: () => <Component {...props} />
+                    },
+                    ...routes
+                ])
+            }
+        )
```
Is there something that we can do to make this easier to read? Coming in after the fact it look a long bit of staring at this code to understand what it's doing.

jeremy-jung1 replied:
If the API trigger was a confusing part, this method of using resolveRef was necessary because the API query is a hook and the only way to update the parameter on that is to update the state. It wouldn't be possible to call the hook within the function although that would make it more readable. The other way would be to do it like in the SPIKE which exposes the access token logic

jeremy-jung1 replied:
I refactored this HOC to break it down into functions. 
[First attempt ](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2288/commits/e95c79e2447caf5e0ac5989019a1dcd343361628#diff-e36f1e70a39b8aa58c63d3ffd83e7430a5d7d5f1463174685e6ccc95e6f4b7b5) I broke it down to functions that are defined within the HOC, and this was good because it meant that a lot of the variables were defined in the same scope and I didn't have to make them into arguments, but downside is it bloated the HOC file.

Since we want to keep the number of functions per file at a minimum, instead I put those functions into a utils file. Hopefully that makes things more readable.

bendvc commented on packages/extension-commerce-bm-seo/src/components/with-optional-commerce-sdk-react-provider.tsx:
```diff
@@ -0,0 +1,77 @@
+import {UserConfig} from '../types/config'
+import {logger} from '../logger'
+
+/**
+ * Checks if the CommerceApiProvider is already installed in the component tree.
+ * @returns boolean, true if the CommerceApiProvider is installed, false otherwise.
+ */
+const useHasCommerceApiProvider = () => {
+    let hasProvider = false
+
+    try {
+        const api = useCommerceApi()
+
+        // the api object is an object with a bunch of api clients like ShopperProduct, ShopperOrder, etc.
+        // if the object is empty, then the CommerceApiProvider is not installed
+        if (Object.keys(api).length > 0) {
+            hasProvider = true
+        }
+    } catch (_) {
+        hasProvider = false
+    }
+
+    return hasProvider
+}
```
Move this to a hooks specific file please, it's possible to also follow the pattern that @hajinsuha1 had moving this to the extension sdk under a commerce folder.

bendvc commented on packages/extension-commerce-bm-seo/src/components/with-optional-commerce-sdk-react-provider.tsx:
```diff
@@ -0,0 +1,77 @@
+    return hasProvider
+}
+
+type WithOptionalCommerceSdkReactProvider = React.ComponentPropsWithoutRef<any>
```
This will probably go with it.

bendvc commented on packages/extension-commerce-bm-seo/src/setup-app.ts:
```diff
@@ -0,0 +1,88 @@
+// Using the `overridable` loader means that you are opting in to the override module resolution flow. As a result this module
+// will be resolved by first looking in the base projects `overrides` folder then the overrides folders of any extensions configured
+// after this one. Only if no module is found will the referenced module in this project be used.
+import seoHOC from 'overridable!./components/seo-hoc'
```
Lets follow standards of using the prefix of `with` for HOC's. And axe the "HOC" post fix. We would want to change the export in that file too if it's names seoHOC and the file name.

jeremy-jung1 committed:
Reapply "template-typescript-minimal setup"

This reverts commit 0f04f9541a336ab6e2c2a469d37c29ac5836c9f3.

yunakim714 committed:
@W-17543649 - [SEO] Before calling `getUrlMapping` check the React Router Config first (#2379)

Added new config value `routingMode` to the SEO extension, which has 2 values, `api_first` or `router_first`. This determines whether the app always calls out to the SEO API, or first checks and matches for routes defined in code.

---------

Signed-off-by: Yuna Kim <84923642+yunakim714@users.noreply.github.com>
Co-authored-by: Jang ho Jung <jangho.jung@salesforce.com>

jeremy-jung1 committed:
Get rid of siteLocale usage and update README a little

jeremy-jung1 committed:
Revert "Reapply "template-typescript-minimal setup""

This reverts commit 9b22815dc91d7d91ea79a217fff80733b8f00e23.

jeremy-jung1 committed:
Apply more feedback

jeremy-jung1 committed:
Merge branch 'useBlock' of github.com:SalesforceCommerceCloud/pwa-kit into useBlock

jeremy-jung1 committed:
Address more feedback

jeremy-jung1 committed:
Update function comment

jeremy-jung1 committed:
Refactoring

jeremy-jung1 committed:
Update apply-application-extensions.test.ts

jeremy-jung1 committed:
linting

jeremy-jung1 committed:
linting

jeremy-jung1 committed:
More refactoring

jeremy-jung1 committed:
Update bootstrap.js

jeremy-jung1 committed:
Update README.md

jeremy-jung1 committed:
Do some ls

jeremy-jung1 committed:
Try removing post install

jeremy-jung1 committed:
Try something

jeremy-jung1 committed:
Update bootstrap.js

jeremy-jung1 committed:
Update bootstrap.js

jeremy-jung1 committed:
Update bootstrap.js

jeremy-jung1 committed:
Update HOC wording in test

vmarta commented on packages/extension-commerce-bm-seo/package.json:
```diff
@@ -1,40 +1,37 @@
+    "@salesforce/pwa-kit-extension-sdk": "4.0.0-extensibility-preview.5",
+    "@salesforce/pwa-kit-react-sdk": "4.0.0-extensibility-preview.5",
+    "@salesforce/pwa-kit-runtime": "4.0.0-extensibility-preview.5",
     "@salesforce/commerce-sdk-react": "^3.2.0-extensibility-preview.4",
```
I cannot build/run this PR via `npm ci` at the root. Seeing this error, which is also similar to what's seen in CI:
```
erna ERR! lerna ENOENT: no such file or directory, open '/Users/vmarta/repos/pwa-kit/packages/extension-commerce-bm-seo/node_modules/.bin/pwa-kit-extension-sdk'
node:child_process:957
    throw err;
    ^
```

Could it be because of the changes in this package.json file?

For example, try updating this version number, since `preview.5` is the latest version for it. 
```suggestion
    "@salesforce/commerce-sdk-react": "^3.2.0-extensibility-preview.5",
```

It was the older `preview.4` because it used to be under the `peerDependencies`. @kevinxh was saying that it's a known issue with our monorepo... the peer dep should be a version that's already published to npm.
